src/parser/lexer.mll

```diff
@@ -692,24 +727,77 @@ and bad_markup_recovery start_offset input = parse
- | (space_char+ as prefix)
-   ((space_char* (_ # space_char # ['['])+)+ as meta)
-   ((space_char* '[') as suffix)
-    {
-      let meta =
-        with_location_adjustments ~adjust_start_by:prefix ~adjust_end_by:suffix (fun _ -> Loc.at) input meta
-      in
-      `Ok (Some meta)
-    }
-  | (newline | horizontal_space)* '['
-    { `Ok None }
-  | _ as c
-    { `Invalid_char c }
+(* Based on OCaml's parsing/lexer.mll
+   We're missing a bunch of cases here, and can add them
+   if necessary. Using the missing cases will cause a warning *)
+and string input = parse
+ | '\"'
+   { let result = Buffer.contents string_buffer in
+     Buffer.clear string_buffer;
+     result }
+ | '\\' newline [' ' '\t']*
+   { string input lexbuf }
+ | '\\' (['\\' '\'' '\"' 'n' 't' 'b' 'r' ' '] as c)
+   { Buffer.add_char string_buffer
+       (match c with
+        | '\\' -> '\\'
+        | '\'' -> '\''
+        | '\"' -> '\"'
+        | 'n' -> '\n'
+        | 't' -> '\t'
+        | 'b' -> '\b'
+        | 'r' -> '\r'
+        | ' ' -> ' '
+        | _ -> assert false);
+     string input lexbuf }
+  | '\\' ['0'-'9'] ['0'-'9'] ['0'-'9']
+    { Buffer.add_char string_buffer (char_for_decimal_code input lexbuf 1);
+      string input lexbuf }
+  | '\\' (_ as c)
+    { warning input (Parse_error.should_not_be_escaped c);
+      Buffer.add_char string_buffer c;
+      string input lexbuf }
   | eof
+    { warning input Parse_error.truncated_string;
+      Buffer.contents string_buffer }
+  | (_ as c)
+    { Buffer.add_char string_buffer c;
+      string input lexbuf }
+  
+and code_block_metadata_atom input = parse
+ | '"'
+   {
+    let start_offset = Lexing.lexeme_start input.lexbuf + 1 in
+    Buffer.clear string_buffer;
+    let s = string input lexbuf in
+    with_location_adjustments ~start_offset ~adjust_end_by:"\"" (fun _ -> Loc.at) input s }
+ | (tag_unquoted_atom as value)
+   { with_location_adjustments (fun _ -> Loc.at) input value }
+ | (_ as c) 
+   { warning input (Parse_error.code_block_tag_invalid_char c);
+     with_location_adjustments (fun _ -> Loc.at) input "" }
+
+and code_block_metadata_tail input tag acc = parse
+ | space_char+
+   { let acc = match tag with | Some t -> `Tag t :: acc | None -> acc in
+     let tag = code_block_metadata_atom input lexbuf in
+     code_block_metadata_tail input (Some tag) acc lexbuf }
+ | space_char* '[' (* Nb this will be a longer match than the above case! *)
+   {
+     let acc = match tag with | Some t -> `Tag t :: acc | None -> acc in
+     `Ok (List.rev acc) }
+ | '='
+   { match tag with
+     | Some t ->
+       let value = code_block_metadata_atom input lexbuf in
+       code_block_metadata_tail input None (`Binding (t, value) :: acc) lexbuf
+     | None ->
+       warning input (Parse_error.code_block_tag_invalid_char '=');
+       code_block_metadata_tail input None acc lexbuf }
+ | (_ # space_char # '[' # '=' as c) [^ '[']* '['
+   { warning input (Parse_error.code_block_tag_invalid_char c);
```

Comment:
In this case we are omitting all the rest of the tags. Might be worth mentioning in the error message, as one might be surprised that:
```
{@ocaml x=y \" z=t[
foo
]}
```
has binding `x=y` but not `z=t`.

Reply:
(or maybe the invalid character could be ignored, just like invalid `=` are ignored?

Reply:
It's a little tricky because this bit of the lexer is required to start with whitespace, which makes the errors a bit odd. If we just skip the char, then for example, the test `{@ocaml =foo [ code ]}` if would complain about the '=' and then it will also complain about the 'f', the 'o' and the 'o'. It's currently trying to skip the rest of the metadata until the '[', but I could try to make it just skip until some whitespace.